protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,230 @@
+# Allowing Non-Determinism in `ProcessProposal` (a.k.a. addressing #1174)
+
+## Context
+
+### Consensus Properties
+
+Byzantine Fault Tolerant (BFT) Consensus is usually specified by the following properties (2 invariants and the liveness property termination):
+
+- _agreement_: no two correct processes decide differently.
+- _validity_: the decided block `v` satisfies the predefined predicate `valid(v)`.
+- _termination_: all correct processes eventually decide.
+
+The consensus algorithm implemented in CometBFT (Tendermint) guarantees these properties for each height.
+
+> To be a bit more formal, valid it a mathematical function here that depends only on its inputs.
+> The arxiv paper glosses over some details here. In fact, valid is a function of the proposed block `v`
+> and the state current state of the blockchain after applying the block of height `h-1`; let's write it as
+> bc(h-1). This we may write `valid(v, bc(h-1))`. Observe that this is a "mathematical function" that is
+> it only depends on its input. (In contrast to functions in progamming that may use other data to 
+> compute the result.
+
+### ABCI interface
+
+In the version of ABCI (`v0.17.0`) that existed before ABCI 1.0 and 2.0 (a.k.a. ABCI++),
+the implementation of the `valid(v)` predicate was totally internal to CometBFT.
+The application had no direct say on the validity of a block,
+although it could (and still can) influence it indirectly via the best-effort ABCI call `CheckTx`.
+
+With the evolution of ABCI to ABCI 1.0 and 2.0, the `valid(v)` predicate now has two components:
+
+- the validity checks performed directly by CometBFT on blocks
```

Comment:
In an abstract wording, can `v` be appended to the blockchain, thus becoming the block at the current height `h`?

Reply:
We could, but I don't see that is needed for this write-up